ENH: sparse: Add DOK support for 1d (without indexing)

dschult opened the pull request:
This PR adds DOK support for 1d with a small footprint to ease review.
I've removed any support for indexing beyond integer indices. That will be a separate PR.

The new test file `test_common1d.py` has tests taken from `test_base.py` revamped for 1d. In this PR many are skipped as not relevant for this limited dok 1d support. I can remove the skipped tests if that is preferred. Similarly I could move away from test classes toward parametrized functions if that is preferred.

Notes:
- _`getcol` and `_getrow` are still returning 2d row/col objects -- as stated in the doc_string.  It's not clear what to return in the 1d case for these methods. And I'm not sure what the long term plan is for `getrow` and `getcol` or the relationship between them and their private partners discussed here. 
- `__add__` currently converted to `csc` and adds when the format does not match. Changing to `coo` and then using that datastructure to construct a dok result is implemented here. I've left commented out code that could handle 1d separately from 2d and keep the old `csc` behavior.  I need to change this to remove the comments, or remove the tocoo approach. 
- `__radd__` is a duplicate of `__add__` so why keep it? Please check if I've missed something -- especially if we should leave it. I left it as just calling `__add__` in case there is a subtlety that occurs if you remove it.
- `resize` works within 1d and within 2d. But raises an exception if asked to resize from 1d to 2d or vice versa.
- `astype` on `_spbase` uses `tocsr` so I added an overwrite version to make it work without converting. This works for 1d or 2d. So as written it changes the code run for 2d to convert the dtype of the dok array. I could call the super() version for 2d if preferred.

This PR builds on #18530 and that should be merged first.

Feedback is welcome.

dschult committed:
add dok-1d support without minmax or index; enable test GetSet

rgommers commented:
Thanks @dschult, looks promising! I won't have time for a full review before the holidays, but wanted to answer this:

> The new test file `test_common1d.py` has tests taken from `test_base.py` revamped for 1d. In this PR many are skipped as not relevant for this limited dok 1d support. I can remove the skipped tests if that is preferred. Similarly I could move away from test classes toward parametrized functions if that is preferred.

If you expect the skipped tests to become useful to a future PR, where you add functionality and remove the skip, then it seems fine to me to leave them in now. If they will remain unused, then it seems better to remove them.

Test classes vs. parametrized functions: there's not enough difference I think to justify rework. For completely new code we tend to prefer parametrized functions, but if this works now and you want to leave it as is, that seems perfectly fine to me (especially since it matches what other 2-D tests do).

dschult commented:
I've updated this PR to modernize the tests (which were taken from test_base) and to remove tests that aren't relevant to the `dok_array` format. I will add those other tests when we add the other functionality.  The new tests are now less than 500 lines -- and are more readable, so hopefully that helps with review.

Two other larger questions that I've identified:
- in `dok_array.__add__` it used to convert to `csc` to do the addition. This version converts to `coo` to do the addition. That works for 1d and 2d and seems faster in my very crude tests. I think the only change in values will be round-off due to different order of operations. It passes all the tests. The question is whether I should make it use the `csc` route for 2d and the coo route for 1d, or is it OK to change the 2d to the `coo` route.
- there doesn't appear to be a `dok_array.eliminate_zeros` method and there probably should be. I have added it as a commented out function with a comment-question in this PR. The question is whether we should include this while we're upgrading dok anyway.

perimosocordiae commented on scipy/sparse/_base.py:
```diff
@@ -867,29 +871,36 @@ def _getcol(self, j):
+        if self.ndim == 1:
+            # only here to force col to be 2d ("as an (mx1) sparse array")
+            # should this be 1d? -- hard to know with name "getcol"
+            return (self @ col_selector).reshape(1, 1)
```
I'd prefer to do:

```
result = self @ col_selector
```

before the `ndim` check.

perimosocordiae commented on scipy/sparse/_base.py:
```diff
@@ -867,29 +871,36 @@ def _getcol(self, j):
         """
+        if self.ndim == 1:
+            # force col to be 2d ("as an (1xn) sparse array")
+            return self.reshape(1, self.shape[-1])
```
I think it makes sense to throw an error here, as `getrow()` isn't really meaningful for a 1d container.

perimosocordiae commented on scipy/sparse/_base.py:
```diff
@@ -1177,6 +1196,12 @@ def _is_integral(dtype):
+        if self.ndim == 1:
+            if axis not in (None, -1, 0):
+                raise ValueError("axis must be None, -1 or 0")
+            res = inter_self / np.array(self.shape[-1])
```
Why is the `np.array()` call needed here?

dschult replied:
`np.array` is used for 2d, so I kept it in 1d. But removing it in both makes more sense and passes all tests.

I have made a commit to put in all these suggestions and also remove the extra `np.array` from the 2d code 3-4 lines down from this section. Remaining questions about these 3 suggestions are:
- should we raise error in `getcol` for 1d?
- should we remove the `np.array` from the 2d code?

dschult committed:
update for indices->coords

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -22,50 +22,61 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
-            if arg1.format == self.format and copy:
-                arg1 = arg1.copy()
+            if arg1.format == self.format:
+                if copy:
```
Discussed clarifying semantics

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -22,50 +22,61 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
 
-            self._dict.update(arg1)
-            self._shape = check_shape(arg1.shape)
+            self._update(arg1._dict)
```
Can dict be set directly from arg1?

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -22,50 +22,61 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
+            if arg1.ndim == 1:
+                if dtype is not None:
+                    arg1 = arg1.astype(dtype)
+                self._update((i, v) for i, v in enumerate(arg1.ravel()) if v != 0)
```
Set dict directly using dict comprehension?

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -22,50 +22,61 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
+                self.dtype = arg1.dtype
+            else:
+                d = self._coo_container(arg1, dtype=dtype).todok()
+                self._update(d._dict)
```
Set dict directly?

stefanv replied:
Consider removing `_update`

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -207,63 +318,50 @@ def __add__(self, other):
-            M, N = self.shape
-            for key in itertools.product(range(M), range(N)):
+            # Add this scalar to each element.
+            for key in itertools.product(*map(range, self.shape)):
```
```
(range(d) for d in self.shape)
```

?

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -207,63 +318,50 @@ def __add__(self, other):
+                raise ValueError("Matrix dimensions are not equal.")
+            res_dtype = upcast(self.dtype, other.dtype)
+            new = self._dok_container(self.shape, dtype=res_dtype)
+            new._dict.update(self._dict)
```
set dict? or set dict with copy?

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -207,63 +318,50 @@ def __add__(self, other):
+                    new._dict.update((k, new[k] + v) for k, v in other.items())
+            else:  # elif self.ndim == 1:
+                # this works for 2d too, but maybe subtle differences?
+                o_coo = other.tocoo()
```
This implementation looks better to me.

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -274,19 +372,55 @@ def _mul_scalar(self, other):
-        # matrix * vector
-        result = np.zeros(self.shape[0], dtype=upcast(self.dtype, other.dtype))
+        res_dtype = upcast(self.dtype, other.dtype)
+
```
Double check multiplication semantics

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -274,19 +372,55 @@ def _mul_scalar(self, other):
-        for (i, j), v in self.items():
-            result[i,:] += v * other[j,:]
+        # vector @ multivector
+        if self.ndim == 1:
```
Double check multiplication semantics; align with CSR / array API as necessary

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -274,19 +372,55 @@ def _mul_scalar(self, other):
+        # vector @ multivector
+        if self.ndim == 1:
+            if other.ndim == 1:
+                result_shape = ()
```
Return scalar?

dschult replied:
To follow array api, return a 0d array with the dot product as the only value.

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -368,23 +516,54 @@ def todok(self, copy=False):
+            # TODO implement resize across dimensions
+            raise NotImplementedError
+
+        if self.ndim == 1:
```
Recommend not implementing resize semantics; even with NumPy, its behavior is  inconsistent.

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -368,23 +516,54 @@ def todok(self, copy=False):
+        if self.dtype != dtype:
+            result = self._dok_container(self.shape, dtype=dtype)
+            data = np.array(list(self._dict.values()), dtype=dtype)
+            result._update(zip(self._dict, data))
```
Update or assign?

stefanv reviewed (approved):
Pending comments, which are minor.

dschult committed:
remove _update to assign to _dict. simplify and clarify code as per review

dschult reviewed (commented):
Changes after review:
- switched from using `_update` to setting `._dict` attribute. 
- `_update` method now longer used so removed.
- improved readability of code and logic
- simplified `_matmul_multivector` code and aligned result shapes with CSR and array api

stefanv commented:
This looks great, Dan! What did you decide to do about `resize`?

dschult commented:
For `resize` I didn't make any changes yet -- I left that discussion unresolved because I think I need more research. Any changes that are made should affect all formats and 1d & 2d ndims. So it might be better as a separate PR.

`sp.sparse` is very clear in the docs that this `resize` is NOT like either of numpy's `resize`. Numpy has different behavior for `numpy.resize` and `numpy.ndarray.resize`. The [docs for `numpy.resize`](https://numpy.org/doc/stable/reference/generated/numpy.resize.html) seem to discourage its use by saying if the number of elements is the same, use `reshape`, and otherwise use indexing or padding. And it has a warning that axes are not considered when placing elements in the array. Those are lots of reasons to look elsewhere. The [docs for `numpy.ndarray.resize`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.resize.html#numpy.ndarray.resize) explain that this method does three things: flatten then resize in 1d then reshape. 

The docs for `scipy.sparse.*_array.resize` clearly make this method different from numpy.
```
The semantics are not identical to [numpy.ndarray.resize](https://numpy.org/devdocs/reference/generated/numpy.ndarray.resize.html#numpy.ndarray.resize) or [numpy.resize](https://numpy.org/devdocs/reference/generated/numpy.resize.html#numpy.resize). Here, the same data will be maintained at each index before and after reshape, if that index is within the new bounds. In numpy, resizing maintains contiguity of the array, moving elements around in the logical array but not within a flattened representation.
```
But it is not clear why this functionality cannot be replaced with indexing and padding.

This api makes no attempt to match the behavior of Numpy.  And there is no Array API to match (no `resize`). So we really shouldn't be using the name `resize` here to avoid confusion with numpy's behavior.

I recommend:
- we deprecate `resize` for sparse arrays (not matrices).
- make sure that indexing and padding can handle all behavior of `resize` and write that up somehow.
- create other-named methods to provide any `resize`-like behavior not possible with indexing and padding.

But, those changes seem sufficiently different from this PR in intent and code coverage (one method for many formats, instead of many methods for one format) that I'd like to try it in a different PR.  Discussion here makes sense though, before I try to implement anything. 

Suggestions?

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -207,63 +310,43 @@ def __add__(self, other):
-            M, N = self.shape
-            for key in itertools.product(range(M), range(N)):
+            # Add this scalar to each element.
+            for key in itertools.product(*[range(d) for d in self.shape]):
```
This is only for 2d, right? So we don't need to make it generic to nd shapes.

dschult replied:
Hard to tell from the diff, but this is in `__add__`, not the `_set_*` methods. So it needs to be 2d or 1d (or eventually maybe nd).

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -101,27 +103,80 @@ def keys(self):
+        if key in self._dict:
+            return self._dict.get(key, default)
+        if isintlike(key) and self.ndim == 1:
+            key = (key,)
```
I'm not sure we want to do this, because the 1d DOK format stores its keys as plain integers.

We might need a fully separate codepath for the 1d case in this function.

dschult replied:
I forgot to unwrap the key in the 1d case before the final line.

Also, this now allows negative indices for DOK in `get`. That brings it closer to `A[key]`.

I can remove the handling of negative indices -- and can make a separate path for 1d if you'd prefer.

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -101,27 +103,80 @@ def keys(self):
+    def _get_int(self, idx):
+        return self._dict.get(idx, self.dtype.type(0))
+
+# Only supporting integer indexing right now. Soon we will use 1d slice and array.
```
Let's remove all this commented code for now. We can keep it around in a separate branch.

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -207,63 +310,43 @@ def __add__(self, other):
-                new._dict.update(self._dict)
                 with np.errstate(over='ignore'):
-                    new._dict.update((k, new[k] + other[k]) for k in other.keys())
+                    new._dict.update((k, new[k] + v) for k, v in other.items())
```
Let's unify this with the version below:

```python
if other.format == "dok":
  o_items = other.items()
else:
  other = other.tocoo()
  if self.ndim == 1:
    o_items = zip(other.coords[0], other.data)
  else:
    o_items = zip(zip(*other.coords), other.data)
```

dschult replied:
Nice!

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -274,19 +357,40 @@ def _mul_scalar(self, other):
+                    shared_keys = self.keys() & other.keys()
+                else:
+                    shared_keys = self.keys() & other.tocoo().coords[0]
+                arr = np.array([self._dict[k] * other._dict[k] for k in shared_keys])
```
Can we do: `sum(self._dict[k] * other._dict[k] for k in shared_keys)` and save some temporary allocations?

dschult committed:
review suggestions

dschult committed:
remove 'dok_array' from exception message

perimosocordiae commented:
Merged! 🎉
